core/metacling/src/TCling.cxx

```diff
@@ -1122,7 +1122,59 @@ static GlobalModuleIndex *loadGlobalModuleIndex(SourceLocation TriggerLoc, cling
          RecreateIndex = true;
       }
       if (RecreateIndex) {
-         GlobalModuleIndex::writeIndex(CI.getFileManager(), CI.getPCHContainerReader(), ModuleIndexPath);
+         cling::Interpreter::PushTransactionRAII deserRAII(&interp);
+         clang::GlobalModuleIndex::UserDefinedInterestingIDs IDs;
+
+         struct DefinitionFinder : public RecursiveASTVisitor<DefinitionFinder> {
+            DefinitionFinder(clang::GlobalModuleIndex::UserDefinedInterestingIDs& IDs,
+                             clang::TranslationUnitDecl* TU) : DefinitionIDs(IDs) {
+               TraverseDecl(TU);
+            }
+            bool VisitNamedDecl(NamedDecl *ND) {
+               for (auto R : ND->redecls()) {
+                  if (!R->isFromASTFile())
+                     continue;
+                  if (TagDecl *TD = llvm::dyn_cast<TagDecl>(R)) {
+                     if (TD->isCompleteDefinition())
+                        Register(TD);
+                  } else if (NamespaceDecl *NSD = llvm::dyn_cast<NamespaceDecl>(R))
+                     Register(NSD, /*AddSingleEntry=*/ false);
+                  else if (TypedefNameDecl *TND = dyn_cast<TypedefNameDecl>(R))
+                     Register(TND);
+                  // FIXME: Add the rest...
+               }
+               return true; // continue decending
+            }
+         private:
+            clang::GlobalModuleIndex::UserDefinedInterestingIDs &DefinitionIDs;
+            void Register(NamedDecl* ND, bool AddSingleEntry = true) {
+               assert(ND->isFromASTFile());
+               if (!ND->hasOwningModule()) {
+                  SourceManager &SM = ND->getASTContext().getSourceManager();
+                  SourceLocation Loc = ND->getLocation();
+                  const FileEntry *FE = SM.getFileEntryForID(SM.getFileID(Loc));
```

Comment:
Is that statement just for the purpose of the assert or does it have another function.
If "only for assert" then maybe put it inside the assert
if "for more", could you add the explanation as a comment (other another maintainer might factor it out).

Reply:
I prefer shorter statements. The optimizer will kill that code anyway. 

Btw, this is a temporary code as if we implement what we discussed to fix the CMSSW non-defining but composing Linkdef by updating the owning module of the declarations that came from the lookup helper and other operations all decls will have an owning module and this code will go.

Reply:
I strongly recommend you put both statements ("code has no side effect and is here just for the assert" and "will be removed later because .... ") in comments.

Reply:
Sure.